search/pre_ranker.cpp

```diff
@@ -147,111 +143,80 @@ void PreRanker::FillMissingFieldsInPreResults()
-    // features and then select a subset with smallest values of this
-    // hash.  In this case this subset of results will be persistent
-    // to small changes in the original set.
-    shuffle(b, e, m_rng);
-  }
 
-  set<PreRankerResult, LessFeatureID> filtered;
+  if (m_results.size() <= BatchSize())
+    return;
+
+  vector<size_t> indices(m_results.size());
+  generate(indices.begin(), indices.end(), [n = 0] () mutable { return n++; });
+  unordered_set<size_t> filtered;
 
-  auto const numResults = min(m_results.size(), BatchSize());
-  filtered.insert(m_results.begin(), m_results.begin() + numResults);
+  auto const iBeg = indices.begin();
+  auto const iMiddle = iBeg + BatchSize();
+  auto const iEnd = indices.end();
 
-  if (!viewportSearch)
+  nth_element(iBeg, iMiddle, iEnd, CompareIndices(&PreRankerResult::LessDistance, m_results));
+  filtered.insert(iBeg, iMiddle);
+
+  if (!m_params.m_categorialRequest)
   {
-    if (!m_params.m_categorialRequest)
-    {
-      nth_element(m_results.begin(), m_results.begin() + numResults, m_results.end(),
-                  &PreRankerResult::LessRankAndPopularity);
-      filtered.insert(m_results.begin(), m_results.begin() + numResults);
-      nth_element(m_results.begin(), m_results.begin() + numResults, m_results.end(),
-                  &PreRankerResult::LessByExactMatch);
-      filtered.insert(m_results.begin(), m_results.begin() + numResults);
-    }
-    else
-    {
-      double const kPedestrianRadiusMeters = 2500.0;
-      PreRankerResult::CategoriesComparator comparator;
-      comparator.m_positionIsInsideViewport =
-          m_params.m_position && m_params.m_viewport.IsPointInside(*m_params.m_position);
-      comparator.m_detailedScale = mercator::DistanceOnEarth(m_params.m_viewport.LeftTop(),
-                                                             m_params.m_viewport.RightBottom()) <
-                                   2 * kPedestrianRadiusMeters;
-      comparator.m_viewport = m_params.m_viewport;
-      nth_element(m_results.begin(), m_results.begin() + numResults, m_results.end(), comparator);
-      filtered.insert(m_results.begin(), m_results.begin() + numResults);
-    }
+    nth_element(iBeg, iMiddle, iEnd, CompareIndices(&PreRankerResult::LessRankAndPopularity, m_results));
+    filtered.insert(iBeg, iMiddle);
+    nth_element(iBeg, iMiddle, iEnd, CompareIndices(&PreRankerResult::LessByExactMatch, m_results));
+    filtered.insert(iBeg, iMiddle);
+  }
+  else
+  {
+    double constexpr kPedestrianRadiusMeters = 2500.0;
+    PreRankerResult::CategoriesComparator comparator;
+    comparator.m_positionIsInsideViewport =
+        m_params.m_position && m_params.m_viewport.IsPointInside(*m_params.m_position);
+    comparator.m_detailedScale = mercator::DistanceOnEarth(m_params.m_viewport.LeftTop(),
+                                                           m_params.m_viewport.RightBottom()) <
+                                 2 * kPedestrianRadiusMeters;
```

Comment:
Не заметил вторую скобку в конце.

```suggestion
    comparator.m_detailedScale = 2 * kPedestrianRadiusMeters >
        mercator::DistanceOnEarth(m_params.m_viewport.LeftTop(), m_params.m_viewport.RightBottom());
```

Reply:
тут не понял

Reply:
Кривое форматирование было.